apis/core/v1beta1/function_webhook.go

```diff
@@ -38,4 +43,418 @@ var _ webhook.Defaulter = &Function{}
+var _ webhook.Validator = &Function{}
+
+func (r *Function) ValidateCreate() error {
+	functionlog.Info("validate create", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateUpdate(old runtime.Object) error {
+	functionlog.Info("validate update", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateDelete() error {
+	functionlog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+func (r *Function) Validate() error {
+	if r.Spec.Image == "" {
+		return errors.New("spec.image must be set")
+	}
+
+	if r.Spec.Build != nil {
+		if r.Spec.ImageCredentials == nil {
+			return errors.New("spec.imageCredentials must be set when `spec.build` is enabled")
+		}
+		if r.Spec.ImageCredentials != nil && r.Spec.ImageCredentials.Name == "" {
+			return errors.New("spec.imageCredentials.name must be set when `spec.build` is enabled")
+		}
+		if err := r.ValidateBuild(); err != nil {
+			return err
+		}
+	}
+
+	if r.Spec.Serving != nil {
+		if err := r.ValidateServing(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Function) ValidateBuild() error {
+	if r.Spec.Build.Builder == nil && r.Spec.Build.Dockerfile == nil {
+		return errors.New("one of spec.build.builder and spec.build.dockerfile must exist")
+	}
+
+	if r.Spec.Build.SrcRepo != nil && r.Spec.Build.SrcRepo.Url == "" {
+		return errors.New("spec.build.srcRepo.url cannot be empty string")
+	}
+
+	if r.Spec.Build.Timeout != nil && r.Spec.Build.Timeout.Duration < 0 {
+		return errors.New("spec.build.timeout cannot be less than 0")
+	}
+
+	if r.Spec.Build.SuccessfulBuildsHistoryLimit != nil && *r.Spec.Build.SuccessfulBuildsHistoryLimit < 0 {
+		return errors.New("spec.build.successfulBuildsHistoryLimit cannot be less than 0")
+	}
+
+	if r.Spec.Build.FailedBuildsHistoryLimit != nil && *r.Spec.Build.FailedBuildsHistoryLimit < 0 {
+		return errors.New("spec.build.failedBuildsHistoryLimit cannot be less than 0")
+	}
+
+	if r.Spec.Build.BuilderMaxAge != nil && r.Spec.Build.BuilderMaxAge.Duration < 0 {
+		return errors.New("spec.build.builderMaxAge cannot be less than 0")
+	}
+
+	if r.Spec.Build.Shipwright != nil {
+		if r.Spec.Build.Shipwright.Strategy != nil && r.Spec.Build.Shipwright.Strategy.Kind != nil {
+			strategies := []string{"BuildStrategy", "ClusterBuildStrategy"}
+			if !containsString(strategies, *r.Spec.Build.Shipwright.Strategy.Kind) {
+				return errors.New("spec.build.shipwright.strategy.kind available values are \"BuildStrategy\", \"ClusterBuildStrategy\"")
+			}
+		}
+
+		if r.Spec.Build.Shipwright.Timeout != nil && r.Spec.Build.Shipwright.Timeout.Duration < 0 {
+			return errors.New("spec.build.shipwright.timeout cannnot be less than 0")
+		}
+	}
+
+	return nil
+}
+
+func (r *Function) ValidateServing() error {
+	if r.Spec.Serving.Runtime == "" {
+		return errors.New("spec.serving.runtime must be set")
+	}
+
+	if !containsString([]string{"knative", "async"}, string(r.Spec.Serving.Runtime)) {
```

Comment:
It is better to refer to the runtime constants: https://github.com/OpenFunction/OpenFunction/blob/main/apis/core/v1beta1/serving_types.go#L30-L31